flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java

```diff
@@ -615,4 +632,85 @@ protected double getCpuCores(final Configuration configuration) {
 		//noinspection NumericCastThatLosesPrecision
 		return cpuCoresLong;
 	}
+
+	/**
+	 * Utility class for converting between Flink {@link WorkerResourceSpec} and Yarn {@link Resource}.
+	 */
+	@VisibleForTesting
+	static class WorkerSpecContainerResourceAdapter {
+		private final Configuration flinkConfig;
+		private final int minMemMB;
+		private final int minVcore;
+		private final boolean matchVcores;
+		private final Map<WorkerResourceSpec, Resource> workerSpecToContainerResource;
+		private final Map<Resource, Collection<WorkerResourceSpec>> containerResourceToWorkerSpecs;
+		private final Map<Integer, Collection<Resource>> containerMemoryToContainerResource;
+
+		@VisibleForTesting
+		WorkerSpecContainerResourceAdapter(
+				final Configuration flinkConfig,
+				final int minMemMB,
+				final int minVcore,
+				final boolean matchVcores) {
+			this.flinkConfig = Preconditions.checkNotNull(flinkConfig);
+			this.minMemMB = minMemMB;
+			this.minVcore = minVcore;
+			this.matchVcores = matchVcores;
+			workerSpecToContainerResource = new HashMap<>();
+			containerResourceToWorkerSpecs = new HashMap<>();
+			containerMemoryToContainerResource = new HashMap<>();
+		}
+
+		@VisibleForTesting
+		Resource getContainerResource(final WorkerResourceSpec workerResourceSpec) {
+			return workerSpecToContainerResource.computeIfAbsent(
+				Preconditions.checkNotNull(workerResourceSpec),
+				this::createAndMapContainerResource);
+		}
+
+		@VisibleForTesting
+		Collection<WorkerResourceSpec> getWorkerSpecs(final Resource containerResource) {
+			return getEquivalentContainerResource(containerResource).stream()
+				.flatMap(resource -> containerResourceToWorkerSpecs.getOrDefault(resource, Collections.emptyList()).stream())
+				.collect(Collectors.toList());
+		}
+
+		@VisibleForTesting
+		Collection<Resource> getEquivalentContainerResource(final Resource containerResource) {
+			// Yarn might ignore the requested vcores, depending on its configurations.
+			// In such cases, we should also not matching vcores.
+			return matchVcores ?
+				Collections.singletonList(containerResource) :
+				containerMemoryToContainerResource.getOrDefault(containerResource.getMemory(), Collections.emptyList());
+		}
+
+		private Resource createAndMapContainerResource(final WorkerResourceSpec workerResourceSpec) {
+			// TODO: need to unset process/flink memory size from configuration if dynamic worker resource is activated
```

Comment:
Same here. I think the RM implementations should not be aware of whether the dynamic worker resources is activated or not.

Reply:
In general I agree unless it is not yet fully supported by the RM implementation. If it is, then I guess we also don't need this comment. If it is not supported, then the check state will help us to remember what needs to be changed.

Reply:
I think after these PRs, the RM implementations should fully support dynamic worker resources. The problem is that we not yet have a SM implementation that supports it. Once we have a SM implementation that supports dynamic worker resources, we should unset the process/flink size from configuration when using it. And if we don't do the unset, Flink might fail at the places of this `TODO`. So I added this `TODO` as a reminder, that if later these codes fail this might be the reason.

Reply:
The current SM will be initialized with the proper default `WorkerResourceSpec` derived from Flink's configuration, right? Hence, we could already resolve this TODO, right? Then there would not be any problems with enabling support for dynamic worker resources.